drivers/network/netio/netio.c

```diff
@@ -0,0 +1,1454 @@
+        }
+        else
+        {
+            s->NextSocketToPut = SocketsToPut;
+            SocketsToPut = s;
+            s->NumSocketPuts = 1;
+        }
+        KeReleaseSpinLock(&SocketsToPutListLock, flags);
+
+        KeSetEvent(&PutSocketsEvent, IO_NO_INCREMENT, FALSE);
+
+        return;
+    }
+
+    s->RefCount--;
+    if (s->RefCount == 0)
+    {
+        SocketShutdown(s);	/* noop when called twice */
+
+        /* Especially listen sockets must keep the LocalAddressFile
+         * open when being closed and there are still accepted sockets
+         * somewhere. Else the accepted socket I/O will fail with
+         * c0000184 (invalid device state). So do not close the
+         * address file in Shutdown only close it here when all AcceptSockets
+         * are gone (AcceptSockets hold a reference to the listen socket).
+         */
+
+        if (s->LocalAddressFile != NULL) {
+            ObDereferenceObject(s->LocalAddressFile);
+            s->LocalAddressFile = NULL;
+        }
+        if (s->LocalAddressHandle != NULL)
+        {
+            ZwClose(s->LocalAddressHandle);
+            s->LocalAddressHandle = NULL;
+        }
+        ExFreePoolWithTag(s, TAG_NETIO);
+    }
+}
+
+NTSTATUS NTAPI
+DriverEntry(PDRIVER_OBJECT DriverObject, PUNICODE_STRING RegistryPath)
+{
+    FUNCTION_TRACE;
+
+    return STATUS_SUCCESS;
+}
+
+static NTSTATUS WSKAPI
+WskSocket(
+    _In_ PWSK_CLIENT Client,
+    _In_ ADDRESS_FAMILY AddressFamily,
+    _In_ USHORT SocketType,
+    _In_ ULONG Protocol,
+    _In_ ULONG Flags,
+    _In_opt_ PVOID SocketContext,
+    _In_opt_ const VOID * Dispatch,
+    _In_opt_ PEPROCESS OwningProcess,
+    _In_opt_ PETHREAD OwningThread,
+    _In_opt_ PSECURITY_DESCRIPTOR SecurityDescriptor,
+    _Inout_ PIRP Irp);
+
+static NTSTATUS NTAPI
+NetioComplete(PDEVICE_OBJECT DeviceObject, PIRP Irp, PVOID Context)
+{
+    struct NetioContext *c = (struct NetioContext *)Context;
+    PIRP UserIrp = c->UserIrp;
+
+    FUNCTION_TRACE;
+
+    if (Irp->IoStatus.Status != STATUS_CANCELLED)
+    {
+        UserIrp->IoStatus.Status = Irp->IoStatus.Status;
+        UserIrp->IoStatus.Information = Irp->IoStatus.Information;
+    }
+
+    if (c->PeerAddrRet != NULL) {
+        PSOCKADDR RemoteAddress =
+            (PSOCKADDR)(&((PTRANSPORT_ADDRESS) c->PeerAddrRet->RemoteAddress)->Address[0].AddressType);
+
+        memcpy(&c->socket->RemoteAddress, RemoteAddress, sizeof(c->socket->RemoteAddress));
+    }
+    if (Irp->IoStatus.Status != STATUS_CANCELLED)
+    {
+        IoCompleteRequest(UserIrp, IO_NETWORK_INCREMENT);
+    }
+
+    SocketPut(c->socket);
+    if (c->TargetConnectionInfo != NULL)
+    {
+        ExFreePoolWithTag(c->TargetConnectionInfo, TAG_AFD_TDI_CONNECTION_INFORMATION);
+    }
+    if (c->PeerAddrRet != NULL)
+    {
+        ExFreePoolWithTag(c->PeerAddrRet, TAG_AFD_TDI_CONNECTION_INFORMATION);
+    }
+    ExFreePoolWithTag(c, TAG_NETIO);
+
+    return STATUS_SUCCESS;
+}
+
+struct ListenContext {
```

Comment:
```suggestion
struct ListenContext
{
```
`+` similar suggestions elsewhere.